server/src/main/java/org/opensearch/common/time/RFC3339DateTimeFormatter.java

```diff
@@ -0,0 +1,422 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.time;
+
+import java.text.ParsePosition;
+import java.time.DateTimeException;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeParseException;
+import java.time.temporal.TemporalAccessor;
+import java.util.Arrays;
+import java.util.Locale;
+
+/**
+ * Defines a close profile of RFC3339 datetime format where the date is mandatory and the time is optional.
+ * <p>
+ * The returned formatter can only be used for parsing, printing is unsupported.
+ * <p>
+ * This parser can parse zoned datetimes.
+ * The parser is strict by default, thus time string {@code 24:00} cannot be parsed.
+ * <p>
+ * It accepts formats described by the following syntax:
+ * <pre>
+ * Year:
+ *       YYYY (eg 1997)
+ *    Year and month:
+ *       YYYY-MM (eg 1997-07)
+ *    Complete date:
+ *       YYYY-MM-DD (eg 1997-07-16)
+ *    Complete date plus hours and minutes:
+ *       YYYY-MM-DDThh:mmTZD (eg 1997-07-16T19:20+01:00)
+ *    Complete date plus hours, minutes and seconds:
+ *       YYYY-MM-DDThh:mm:ssTZD (eg 1997-07-16T19:20:30+01:00)
+ *    Complete date plus hours, minutes, seconds and a decimal fraction of a second
+ *       YYYY-MM-DDThh:mm:ss.sTZD (eg 1997-07-16T19:20:30.45+01:00)
+ *       YYYY-MM-DDThh:mm:ss,sTZD (eg 1997-07-16T19:20:30,45+01:00)
+ * where:
+ *
+ *      YYYY = four-digit year
+ *      MM   = two-digit month (01=January, etc.)
+ *      DD   = two-digit day of month (01 through 31)
+ *      hh   = two digits of hour (00 through 23) (am/pm NOT allowed)
+ *      mm   = two digits of minute (00 through 59)
+ *      ss   = two digits of second (00 through 59)
+ *      s    = one or more(max 9) digits representing a decimal fraction of a second
+ *      TZD  = time zone designator (Z or z or +hh:mm or -hh:mm)
+ * </pre>
+ */
+final class RFC3339DateTimeFormatter extends OpenSearchDateTimeFormatter {
```

Comment:
The parser is supporting RFC3339 and more (i.e. lenient in nature) based on my reading of the code and RFC. @reta @CaptainDredge Do you think we should rename it to `rfc3339_lenient` instead so that users are more clear on the validations/support?

Reply:
If introducing a new parser/formatter, why not make it according to the specification? This is more than a little confusing.

Reply:
Also this code is stolen from https://github.com/ethlo/itu/blob/v1.7.2/src/main/java/com/ethlo/time/internal/EthloITU.java

Reply:
> If introducing a new parser/formatter, why not make it according to the specification? This is more than a little confusing

We want to support lenient formats because of wider usability.  For eg. [in this PR](https://github.com/opensearch-project/OpenSearch/pull/11675) we'll try to add another format to support local datetime which is also a profile under ISO8601 and will be helpful for most popular logging library datetime formats

Reply:
> Also this code is stolen from https://github.com/ethlo/itu/blob/v1.7.2/src/main/java/com/ethlo/time/internal/EthloITU.java

@CaptainDredge it this the case? If yes - the license is permissive - please include the attribution / reference to original code location 

@ethlo sincere apologies for that (if it is the case)